src/io/flutter/jxbrowser/JxBrowserManager.java

```diff
@@ -0,0 +1,209 @@
+  private static JxBrowserManager manager;
+  private static final String DOWNLOAD_PATH = FileUtils.platformPath();
+  private static final AtomicReference<JxBrowserStatus> status = new AtomicReference<>(JxBrowserStatus.NOT_INSTALLED);
+  private static final Logger LOG = Logger.getInstance(JxBrowserManager.class);
+  // We will be gating JxBrowser features until all of the features are landed.
+  // To test JxBrowser, set this to true and also add license key to VM options (-Djxbrowser.license.key=<key>).
+  public static final boolean ENABLE_JX_BROWSER = false;
+
+  private JxBrowserManager() {}
+
+  public static JxBrowserManager get() {
+    if (manager == null) {
+      return new JxBrowserManager();
+    }
+    return manager;
+  }
+
+  public boolean isInstalled() {
+    return status.get().equals(JxBrowserStatus.INSTALLED);
+  }
+
+  public void setUp(Project project) {
+    if (!status.compareAndSet(JxBrowserStatus.NOT_INSTALLED, JxBrowserStatus.INSTALLATION_IN_PROGRESS)) {
+      // This check ensures that an IDE only downloads and installs JxBrowser once, even if multiple projects are open.
+      // If already in progress, let calling point wait until success or failure (it may make sense to call setUp but proceed).
+      // If already succeeded or failed, no need to continue.
+      return;
+    }
+
+    LOG.info(project.getName() + ": Installing JxBrowser");
+
+    final File directory = new File(DOWNLOAD_PATH);
+    if (!directory.exists()) {
+      if (!directory.mkdirs()) {
+        LOG.info(project.getName() + ": Unable to create directory for JxBrowser files");
+        status.set(JxBrowserStatus.INSTALLATION_FAILED);
+        return;
+      }
+    }
+
+    // Check for file or loading file
+    final String[] fileNames = {JxBrowserUtils.getPlatformFileName(), JxBrowserUtils.getApiFileName(), JxBrowserUtils.getSwingFileName()};
+    boolean allDownloaded = true;
+    final List<File> files = new ArrayList<>();
+    for (String fileName : fileNames) {
+      final File file = new File(DOWNLOAD_PATH + File.separator + fileName);
+      files.add(file);
+      if (!file.exists()) {
+        allDownloaded = false;
+        break;
+      }
+    }
+
+    if (allDownloaded) {
+      LOG.info(project.getName() + ": JxBrowser platform file already exists, skipping download");
+      loadClasses(files);
+      return;
+    }
+
+    final File tempLoadingFile = new File(DOWNLOAD_PATH + File.separator + JxBrowserUtils.getLoadingFileName());
+    boolean created = false;
+    try {
+      created = tempLoadingFile.createNewFile();
+    }
+    catch (IOException e) {
+      e.printStackTrace();
+    }
+
+    if (!created) {
+      // This means another IDE already created this file and has started downloading.
+      // Wait for download to finish and then try loading again.
+      LOG.info(project.getName() + ": Waiting for JxBrowser file to download");
+      int attempts = 0;
+      while (attempts < 100) {
```

Comment:
make 100 a constant and consider removing if using exponential backoff.

Reply:
Do you mean keep increasing the sleep time with more iterations? This would still require some stopping condition, you're thinking something other than number of attempts?

Reply:
yes. what you would do is double the time to sleep each time.
rather than having a stopping condition, what would be ideal would be to have the code that displays the browser check whether the browser is available yet and if it is not available display the following:
1. "Installing JX Browser" (for the common case where JX Browser has not failed to download)
2. JX Browser Download failed (retrying in X seconds) for the case where the download has failed at least once.
This is similar to what you see if you open gmail when offline. Bonus points: add a "retry download now" button.
That way users only see the message about a failed download when they care when they are about to try to use the inspector and it isn't available.

Reply:
Will be adding panel messages in a next change.